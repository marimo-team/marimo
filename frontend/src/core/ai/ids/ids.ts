/* Copyright 2024 Marimo. All rights reserved. */

import type { TypedString } from "@/utils/typed";

export const PROVIDERS = [
  "openai",
  "anthropic",
  "google",
  "ollama",
  "bedrock",
  "deepseek",
  "azure",
] as const;
export type ProviderId = (typeof PROVIDERS)[number];

export type ShortModelId = TypedString<"ShortModelId">;

/**
 * Qualified model id
 * `provider_id/short_model_id`;
 */
export type QualifiedModelId = `${ProviderId}/${ShortModelId}`;

export class AiModelId {
  readonly providerId: ProviderId;
  readonly shortModelId: ShortModelId;

  constructor(providerId: ProviderId, shortModelId: ShortModelId) {
    this.providerId = providerId;
    this.shortModelId = shortModelId;
  }

  get id(): QualifiedModelId {
    return `${this.providerId}/${this.shortModelId}`;
  }

  static parse(id: string): AiModelId {
    if (!id.includes("/")) {
      const providerId = guessProviderId(id);
      return new AiModelId(providerId, id as ShortModelId);
    }

    const [providerId, ...shortModelId] = id.split("/");
    return new AiModelId(
      providerId as ProviderId,
      shortModelId.join("/") as ShortModelId,
    );
  }
}

function guessProviderId(id: string): ProviderId {
  if (id.startsWith("gpt") || id.startsWith("o3") || id.startsWith("o1")) {
    return "openai";
  }
  if (id.startsWith("claude")) {
    return "anthropic";
  }
  if (id.startsWith("gemini") || id.startsWith("google")) {
    return "google";
  }
  if (id.startsWith("deepseek")) {
    return "deepseek";
  }
  return "ollama";
}

export function isKnownAIProvider(providerId: ProviderId): boolean {
  return PROVIDERS.includes(providerId);
}
