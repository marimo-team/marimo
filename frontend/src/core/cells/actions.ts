/* Copyright 2024 Marimo. All rights reserved. */

import { store } from "../state/jotai";
import { Objects } from "@/utils/objects";
import { notebookAtom } from "./cells";
import { scrollAndHighlightCell } from "@/components/editor/links/cell-link";

/**
 * Scroll to the first cell that is currently in "running" state.
 */
export function notebookScrollToRunning() {
  // find cell that is currently in "running" state
  const { cellRuntime } = store.get(notebookAtom);
  const cell = Objects.entries(cellRuntime).find(
    ([cellid, runtimestate]) => runtimestate.status === "running",
  );
  if (!cell) {
    return;
  }
  scrollAndHighlightCell(cell[0], "focus");
}
