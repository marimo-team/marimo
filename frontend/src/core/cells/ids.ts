/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-redeclare */

import { invariant } from "@/utils/invariant";
import { TypedString } from "../../utils/typed";

const lowercase = "abcdefghijklmnopqrstuvwxyz";
const uppercase = lowercase.toUpperCase();
const alphabet = lowercase + uppercase;

/**
 * A typed CellId
 */
export type CellId = TypedString<"CellId">;
export const CellId = {
  /**
   * Create a new CellId, a random 4 letter string.
   */
  create(): CellId {
    let id = "";
    for (let i = 0; i < 4; i++) {
      id += alphabet[Math.floor(Math.random() * alphabet.length)];
    }
    return id as CellId;
  },
};

/**
 * The id of an HTML element that represents a cell
 */
export type HTMLCellId = `cell-${CellId}`;
export const HTMLCellId = {
  /**
   * Create a new HTMLCellId
   */
  create(cellId: CellId): HTMLCellId {
    invariant(cellId != null, "cellId is required");

    return `cell-${cellId}`;
  },
  parse(htmlCellId: HTMLCellId): CellId {
    return htmlCellId.slice(5) as CellId;
  },
  /**
   * Get the cell container ancestor of `element`, if any.
   */
  findElement(element: Element): (Element & { id: HTMLCellId }) | null {
    return element.closest('div[id^="cell-"]');
  },
};

/**
 * A typed UIElementId
 */
export type UIElementId = `${CellId}-${string}`;
export const UIElementId = {
  parse(element: Element): UIElementId | null {
    return element.getAttribute("object-id") as UIElementId | null;
  },
  parseOrThrow(element: Element): UIElementId {
    const id = UIElementId.parse(element);
    invariant(id, "<marimo-ui-element/> missing object-id attribute");
    return id;
  },
};
