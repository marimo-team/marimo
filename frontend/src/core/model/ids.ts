/* Copyright 2023 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-redeclare */

import { invariant } from "@/utils/invariant";
import { TypedString } from "./typed";

/**
 * A typed CellId
 */
export type CellId = TypedString<"CellId">;
let counter = 0;
export const CellId = {
  /**
   * Create a new CellId
   */
  create(): CellId {
    const next = counter++;
    return next.toString() as CellId;
  },

  /**
   * Reset id provider.
   */
  reset() {
    counter = 0;
  },
};

/**
 * The id of an HTML element that represents a cell
 */
export type HTMLCellId = `cell-${CellId}`;
export const HTMLCellId = {
  /**
   * Create a new HTMLCellId
   */
  create(cellId: CellId): HTMLCellId {
    invariant(cellId != null, "cellId is required");

    return `cell-${cellId}` as HTMLCellId;
  },
  parse(htmlCellId: HTMLCellId): CellId {
    return htmlCellId.slice(5) as CellId;
  },
  /**
   * Get the cell container ancestor of `element`, if any.
   */
  findElement(element: Element): (Element & { id: HTMLCellId }) | null {
    return element.closest('div[id^="cell-"]');
  },
};
