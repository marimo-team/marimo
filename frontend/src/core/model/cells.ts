/* Copyright 2023 Marimo. All rights reserved. */
import { createRef } from "react";
import { OutputMessage } from "../kernel/messages";
import { SerializedEditorState } from "../codemirror/types";
import { CellHandle } from "../../editor/Cell";
import { CellId } from "./ids";

export const DEFAULT_CELL_NAME = "__";

/**
 * The status of a cell.
 *
 * queued: queued by the kernel.
 * running: currently executing.
 * idle: not running.
 * stale: stale, likely because the cell was disabled or an ancestor was disabled.
 * disabled-transitively: disabled because an ancestor was disabled.
 */
export type CellStatus =
  | "queued"
  | "running"
  | "idle"
  | "stale"
  | "disabled-transitively";

/**
 * Create a new cell with default state.
 */
export function createCell({
  key,
  ref = createRef(),
  name = DEFAULT_CELL_NAME,
  code = "",
  output = null,
  consoleOutputs = [],
  status = "idle",
  edited = false,
  interrupted = false,
  errored = false,
  stopped = false,
  runElapsedTimeMs = null,
  runStartTimestamp = null,
  lastCodeRun = null,
  serializedEditorState = null,
  config = {},
}: Partial<CellState> & Pick<CellState, "key">): CellState {
  return {
    key: key,
    ref: ref,
    config: config,
    name: name,
    output: output,
    code: code,
    status: status,
    edited: edited,
    interrupted: interrupted,
    errored: errored,
    stopped: stopped,
    runElapsedTimeMs: runElapsedTimeMs,
    runStartTimestamp: runStartTimestamp,
    lastCodeRun: lastCodeRun,
    consoleOutputs: consoleOutputs,
    serializedEditorState: serializedEditorState,
  };
}

export interface CellState {
  /** unique key */
  key: CellId;
  /** user-given name, or default */
  name: string;
  /** current contents of the editor */
  code: string;
  /** a message encoding the cell's output */
  output: OutputMessage | null;
  /** messages encoding the cell's console outputs. */
  consoleOutputs: OutputMessage[];
  /** current status of the cell */
  status: CellStatus;
  /** whether this cell has been modified since its last run */
  edited: boolean;
  /** whether this cell has been interrupted since its last run */
  interrupted: boolean;
  /** whether this cell was stopped with mo.stop */
  stopped: boolean;
  /** snapshot of code that was last run */
  lastCodeRun: string | null;
  /**
   * whether marimo encountered an error when trying to register or run
   * this cell (such as a multiple definition error)
   */
  errored: boolean;
  /** run start time, as seconds since epoch */
  runStartTimestamp: number | null;
  /** run elapsed time, in milliseconds */
  runElapsedTimeMs: number | null;
  /** serialized state of the underyling editor */
  serializedEditorState: SerializedEditorState | null;

  /** handle to access the underlying cell */
  ref: React.RefObject<CellHandle>;

  /** cell configuration */
  config: CellConfig;
}

export interface CellConfig {
  /**
   * If true, the cell and its descendants are unable to run.
   */
  disabled?: boolean;
}
