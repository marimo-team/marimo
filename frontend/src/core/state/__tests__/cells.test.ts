/* Copyright 2023 Marimo. All rights reserved. */
import { beforeEach, describe, expect, it } from "vitest";
import { CellState, createCell } from "../../model/cells";
import { CellsAndHistory, exportedForTesting } from "../cells";
import { CellId } from "@/core/model/ids";

const { initialCellState, reducer, createActions } = exportedForTesting;

function formatCells(cells: CellState[]) {
  return `\n${cells
    .map((cell) => [`key: ${cell.key}`, `code: '${cell.code}'`].join("\n"))
    .join("\n\n")}`;
}

describe("cell reducer", () => {
  let state: CellsAndHistory;
  let firstCellId: CellId;

  const actions = createActions((action) => {
    state = reducer(state, action);
  });

  beforeEach(() => {
    CellId.reset();

    state = initialCellState();
    state.present = [createCell({ key: CellId.create() })];
    firstCellId = state.present[0].key;
  });

  it("can add a cell after another cell", () => {
    actions.createNewCell({
      cellId: firstCellId,
      before: false,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 0
      code: ''

      key: 1
      code: ''"
    `);
  });

  it("can add a cell before another cell", () => {
    actions.createNewCell({
      cellId: firstCellId,
      before: true,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 1
      code: ''

      key: 0
      code: ''"
    `);
  });

  it("can delete a cell", () => {
    actions.createNewCell({
      cellId: firstCellId,
      before: false,
    });
    actions.deleteCell({
      cellId: firstCellId,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 1
      code: ''"
    `);

    // undo
    actions.undoDeleteCell();
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 2
      code: ''

      key: 1
      code: ''"
    `);
  });

  it("can update a cell", () => {
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import numpy as np",
      formattingChange: false,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 0
      code: 'import numpy as np'"
    `);
  });

  it("can move a cell", () => {
    actions.createNewCell({
      cellId: firstCellId,
      before: false,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 0
      code: ''

      key: 1
      code: ''"
    `);

    // move first cell to the end
    actions.moveCell({
      cellId: firstCellId,
      before: false,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 1
      code: ''

      key: 0
      code: ''"
    `);

    // move it back
    actions.moveCell({
      cellId: firstCellId,
      before: true,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 0
      code: ''

      key: 1
      code: ''"
    `);
  });

  it("can drag and drop a cell", () => {
    actions.createNewCell({
      cellId: firstCellId,
      before: false,
    });
    actions.createNewCell({
      cellId: "1" as CellId,
      before: false,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 0
      code: ''

      key: 1
      code: ''

      key: 2
      code: ''"
    `);

    // drag first cell to the end
    actions.dropCellOver({
      cellId: firstCellId,
      overCellId: "2" as CellId,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 1
      code: ''

      key: 2
      code: ''

      key: 0
      code: ''"
    `);

    // drag it back to the middle
    actions.dropCellOver({
      cellId: firstCellId,
      overCellId: "2" as CellId,
    });
    expect(formatCells(state.present)).toMatchInlineSnapshot(`
      "
      key: 1
      code: ''

      key: 0
      code: ''

      key: 2
      code: ''"
    `);
  });

  it("can run cell and receive cell messages", () => {
    // HAPPY PATH
    /////////////////
    // Initial state
    let cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe(null);
    expect(cell.edited).toBe(false);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Update code
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as mo",
      formattingChange: false,
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe(null);
    expect(cell.edited).toBe(true);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Prepare for run
    actions.prepareForRun({
      cellId: firstCellId,
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe("import marimo as mo");
    expect(cell.edited).toBe(false);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Receive queued messages
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: null,
        status: "queued",
        timestamp: new Date(10).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.status).toBe("queued");
    expect(cell.lastCodeRun).toBe("import marimo as mo");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(null);
    expect(cell.runStartTimestamp).toBe(null);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Receive running messages
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: null,
        status: "running",
        timestamp: new Date(20).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.status).toBe("running");
    expect(cell.lastCodeRun).toBe("import marimo as mo");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(null);
    expect(cell.runStartTimestamp).toBe(20);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Console messags shouldn't transition status
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: {
          channel: "console",
          mimetype: "text/plain",
          data: "hello!",
          timestamp: "",
        },
        status: null,
        timestamp: new Date(22).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.status).toBe("running");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(null);
    expect(cell.runStartTimestamp).toBe(20);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Receive output messages
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: {
          channel: "output",
          mimetype: "text/plain",
          data: "ok",
          timestamp: "",
        },
        console: {
          channel: "console",
          mimetype: "text/plain",
          data: "hello!",
          timestamp: "",
        },
        status: "idle",
        timestamp: new Date(33).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(13_000);
    expect(cell.runStartTimestamp).toBe(null);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // EDITING BACK AND FORTH
    /////////////////
    // Update code again
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as mo\nimport numpy",
      formattingChange: false,
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.edited).toBe(true);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Update code should be unedited
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as mo",
      formattingChange: false,
    });
    cell = state.present[0];
    expect(cell.edited).toBe(false);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Update code should be edited again
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as mo\nimport numpy",
      formattingChange: false,
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe("import marimo as mo");
    expect(cell.edited).toBe(true);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // ERROR RESPONSE
    /////////////////
    // Prepare for run
    actions.prepareForRun({
      cellId: firstCellId,
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    // Queue
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: null,
        status: "queued",
        timestamp: new Date(40).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    // Running
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: null,
        status: "running",
        timestamp: new Date(50).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    // Receive error
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: {
          channel: "marimo-error",
          mimetype: "application/vnd.marimo+error",
          data: [
            { type: "exception", exception_type: "ValueError", msg: "Oh no!" },
          ],
          timestamp: "",
        },
        console: null,
        status: "idle",
        timestamp: new Date(61).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe("import marimo as mo\nimport numpy");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(null);
    expect(cell.runStartTimestamp).toBe(null);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // INTERRUPT RESPONSE
    /////////////////
    // Prepare for run
    actions.prepareForRun({
      cellId: firstCellId,
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    // Queue
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: null,
        status: "queued",
        timestamp: new Date(40).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    // Running
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: null,
        console: null,
        status: "running",
        timestamp: new Date(50).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.output).not.toBe(null); // keep old output
    // Receive error
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: {
          channel: "marimo-error",
          mimetype: "application/vnd.marimo+error",
          data: [{ type: "interruption" }],
          timestamp: "",
        },
        console: null,
        status: "idle",
        timestamp: new Date(61).getTime(),
      },
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe(null);
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(11_000);
    expect(cell.runStartTimestamp).toBe(null);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all
  });

  it("can run cell a stale cell", () => {
    // Update code of first cell
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as mo",
      formattingChange: false,
    });
    // Add cell
    actions.createNewCell({
      cellId: firstCellId,
      before: false,
    });
    const secondCell = state.present[1].key;
    // Update code
    actions.updateCellCode({
      code: "mo.slider()",
      cellId: secondCell,
      formattingChange: false,
    });

    // Prepare for run
    actions.prepareForRun({
      cellId: secondCell,
    });

    // Receive queued messages
    actions.handleCellMessage({
      cellId: secondCell,
      message: {
        cell_id: secondCell,
        output: null,
        console: null,
        status: "queued",
        timestamp: new Date(10).getTime(),
      },
    });
    let cell = state.present[1];
    expect(cell.status).toBe("queued");
    expect(cell.lastCodeRun).toBe("mo.slider()");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(null);
    expect(cell.runStartTimestamp).toBe(null);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all

    // Receive stale message
    actions.handleCellMessage({
      cellId: secondCell,
      message: {
        cell_id: secondCell,
        output: null,
        console: null,
        status: "stale",
        timestamp: new Date(20).getTime(),
      },
    });
    cell = state.present[1];
    expect(cell.status).toBe("stale");
    expect(cell.lastCodeRun).toBe("mo.slider()");
    expect(cell.edited).toBe(false);
    expect(cell.runElapsedTimeMs).toBe(null);
    expect(cell.runStartTimestamp).toBe(null);
    expect(cell).toMatchSnapshot(); // snapshot everything as a catch all
  });

  it("can format code and update cell", () => {
    const firstCellId = state.present[0].key;
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as    mo",
      formattingChange: false,
    });
    let cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe(null);
    expect(cell.edited).toBe(true);

    // Run
    actions.prepareForRun({
      cellId: firstCellId,
    });
    actions.handleCellMessage({
      cellId: firstCellId,
      message: {
        cell_id: firstCellId,
        output: {
          channel: "output",
          mimetype: "text/plain",
          data: "ok",
          timestamp: "",
        },
        console: {
          channel: "console",
          mimetype: "text/plain",
          data: "hello!",
          timestamp: "",
        },
        status: "idle",
        timestamp: new Date(33).getTime(),
      },
    });

    // Check steady state
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.edited).toBe(false);
    expect(cell.lastCodeRun).toBe("import marimo as    mo");

    // Format code
    actions.updateCellCode({
      cellId: firstCellId,
      code: "import marimo as mo",
      formattingChange: true,
    });
    cell = state.present[0];
    expect(cell.status).toBe("idle");
    expect(cell.lastCodeRun).toBe("import marimo as mo");
    expect(cell.edited).toBe(false);
  });

  it("can update a cells config", () => {
    const firstCellId = state.present[0].key;
    let cell = state.present[0];
    // Starts empty
    expect(cell.config).toEqual({});

    actions.updateCellConfig({
      cellId: firstCellId,
      config: { autoRun: false },
    });
    cell = state.present[0];
    expect(cell.config.autoRun).toBe(false);

    // Revert
    actions.updateCellConfig({
      cellId: firstCellId,
      config: { autoRun: null },
    });
    cell = state.present[0];
    expect(cell.config.autoRun).toBe(null);
  });
});
