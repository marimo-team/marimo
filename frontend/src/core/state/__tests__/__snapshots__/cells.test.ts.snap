// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cell reducer > can run cell a stale cell 1`] = `
{
  "code": "mo.slider()",
  "config": {},
  "consoleOutputs": [],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "1",
  "lastCodeRun": "mo.slider()",
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "queued",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell a stale cell 2`] = `
{
  "code": "mo.slider()",
  "config": {},
  "consoleOutputs": [],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "1",
  "lastCodeRun": "mo.slider()",
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "stale",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 1`] = `
{
  "code": "",
  "config": {},
  "consoleOutputs": [],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": null,
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 2`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [],
  "edited": true,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": null,
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 3`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 4`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "queued",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 5`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": 20,
  "serializedEditorState": null,
  "status": "running",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 6`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": null,
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": 20,
  "serializedEditorState": null,
  "status": "running",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 7`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": {
    "channel": "output",
    "data": "ok",
    "mimetype": "text/plain",
    "timestamp": "",
  },
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": 13000,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 8`] = `
{
  "code": "import marimo as mo
import numpy",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": true,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": {
    "channel": "output",
    "data": "ok",
    "mimetype": "text/plain",
    "timestamp": "",
  },
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": 13000,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 9`] = `
{
  "code": "import marimo as mo",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": {
    "channel": "output",
    "data": "ok",
    "mimetype": "text/plain",
    "timestamp": "",
  },
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": 13000,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 10`] = `
{
  "code": "import marimo as mo
import numpy",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": true,
  "errored": false,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo",
  "name": "__",
  "output": {
    "channel": "output",
    "data": "ok",
    "mimetype": "text/plain",
    "timestamp": "",
  },
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": 13000,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 11`] = `
{
  "code": "import marimo as mo
import numpy",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": false,
  "errored": true,
  "initialContents": "",
  "interrupted": false,
  "key": "0",
  "lastCodeRun": "import marimo as mo
import numpy",
  "name": "__",
  "output": {
    "channel": "marimo-error",
    "data": [
      {
        "exception_type": "ValueError",
        "msg": "Oh no!",
        "type": "exception",
      },
    ],
    "mimetype": "application/vnd.marimo+error",
    "timestamp": "",
  },
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": null,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;

exports[`cell reducer > can run cell and receive cell messages 12`] = `
{
  "code": "import marimo as mo
import numpy",
  "config": {},
  "consoleOutputs": [
    {
      "channel": "console",
      "data": "hello!hello!",
      "mimetype": "text/plain",
      "timestamp": "",
    },
  ],
  "edited": false,
  "errored": false,
  "initialContents": "",
  "interrupted": true,
  "key": "0",
  "lastCodeRun": null,
  "name": "__",
  "output": {
    "channel": "marimo-error",
    "data": [
      {
        "type": "interruption",
      },
    ],
    "mimetype": "application/vnd.marimo+error",
    "timestamp": "",
  },
  "ref": {
    "current": null,
  },
  "runElapsedTimeMs": 11000,
  "runStartTimestamp": null,
  "serializedEditorState": null,
  "status": "idle",
  "stopped": false,
}
`;
